src/redisearch_rs/rqe_iterators/src/optional.rs

```diff
@@ -0,0 +1,224 @@
+    /// the iterator yields virtual results until [`Optional::max_doc_id`] is reached.
+    ///
+    /// In case child aborts during [`RQEIterator::revalidate`],
+    /// this child is turned into [`None`], changed from the [`Some`] state it starts
+    /// at when creating using [`Optional::new`]. From that point onward all results
+    /// will be virtual until `max_doc_id` is reached.
+    child: Option<I>,
+}
+
+impl<'index, I> Optional<'index, I>
+where
+    I: RQEIterator<'index>,
+{
+    #[inline(always)]
+    /// Creates a new [`Optional`] iterator.
+    ///
+    /// * `max_id` is the inclusive upper bound of document identifiers visited by
+    ///   [`RQEIterator::read`] and [`RQEIterator::skip_to`].
+    /// * `weight` is applied to [`RSIndexResult`] values returned by the
+    ///   child [`RQEIterator`]. When the child is exhausted, the iterator
+    ///   yields virtual [`RSIndexResult`] values without weight until `max_id` is reached.
+    /// * `child` [`RQEIterator`] used and wrapped around by this [`Optional`] iterator
+    pub const fn new(max_id: t_docId, weight: f64, child: I) -> Self {
+        Self {
+            max_doc_id: max_id,
+            weight,
+            result: RSIndexResult::virt()
+                .frequency(1)
+                .field_mask(RS_FIELDMASK_ALL),
+            child: Some(child),
+        }
+    }
+}
+
+impl<'index, I> RQEIterator<'index> for Optional<'index, I>
+where
+    I: RQEIterator<'index>,
+{
+    fn current(&mut self) -> Option<&mut RSIndexResult<'index>> {
+        if let Some(child) = self.child.as_mut()
+            && child.last_doc_id() == self.result.doc_id
+        {
+            child.current()
+        } else {
+            Some(&mut self.result)
+        }
+    }
+
+    fn read(&mut self) -> Result<Option<&mut RSIndexResult<'index>>, RQEIteratorError> {
+        if self.at_eof() {
+            return Ok(None);
+        }
+
+        let maybe_real = self
+            .child
+            .as_mut()
+            .map(|child| {
+                let child_last_doc_id = child.last_doc_id();
+                match child_last_doc_id.cmp(&(self.result.doc_id + 1)) {
+                    cmp::Ordering::Less => child.read(),
+                    cmp::Ordering::Equal => Ok(child.current()),
+                    cmp::Ordering::Greater => Ok(None),
+                }
+            })
+            .transpose()?
+            .flatten();
+
+        self.result.doc_id += 1;
+
+        if let Some(real) = maybe_real {
+            debug_assert!(
+                real.doc_id >= self.result.doc_id,
+                "no backwards reads should be possible"
+            );
+
+            if real.doc_id > self.result.doc_id {
+                // demote to virtual, preserve sequential guarantees
+                return Ok(Some(&mut self.result));
+            }
+
+            real.weight = self.weight;
+            return Ok(Some(real));
+        }
+
+        Ok(Some(&mut self.result))
```

Comment:
Consider merging the `maybe_real.is_none()` and `maybe_real.is_some() && real.doc_id > self.result.doc_id` cases:
```suggestion
            if real.doc_id == self.result.doc_id {
                real.weight = self.weight;
                return Ok(Some(real));
            }
        }

        // demote to virtual, preserve sequential guarantees
        Ok(Some(&mut self.result))
```